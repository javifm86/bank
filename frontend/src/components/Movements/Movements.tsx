interface Movements {
  date: string;
  type: 'withdraw' | 'deposit';
  amount: number;
  balance: number;
}

interface Props {
  movements: Movements[];
}

function Movements({ movements }: Props) {
  const currencyFormatter = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
  });

  return (
    <div className="-mx-4 sm:-mx-8 px-4 sm:px-8 overflow-x-auto">
      <div className="inline-block min-w-full shadow sm:rounded-lg border-b border-gray-200 overflow-hidden">
        <table className="min-w-full divide-y divide-gray-200">
          <thead className="bg-gray-50">
            <tr>
              <th className="px-5 py-3 text-xs font-semibold text-gray-500 uppercase tracking-wider text-center">
                Date
              </th>
              <th className="px-5 py-3 text-xs font-semibold text-gray-500 uppercase tracking-wider text-right">
                Amount
              </th>
              <th className="px-5 py-3 text-xs font-semibold text-gray-500 uppercase tracking-wider text-right">
                Balance
              </th>
            </tr>
          </thead>
          <tbody className="bg-white divide-y divide-gray-200">
            {movements.map(({ amount, balance, date, type }, index) => {
              const isDeposit = type === 'deposit';
              const sign = isDeposit ? '+' : '-';
              const cssAmount = isDeposit ? 'text-green-500' : 'text-red-500';
              const formattedDate = new Date(date).toLocaleDateString('en-US', {
                year: 'numeric',
                month: 'long',
                day: 'numeric',
              });

              return (
                <tr key={index}>
                  <td className="px-5 py-5 text-sm text-center">
                    {formattedDate}
                  </td>
                  <td className={`px-5 py-5 text-sm text-right ${cssAmount}`}>
                    {sign}
                    {currencyFormatter.format(amount)}
                  </td>
                  <td className="px-5 py-5 text-sm text-right">
                    {currencyFormatter.format(balance)}
                  </td>
                </tr>
              );
            })}
          </tbody>
        </table>
      </div>
    </div>
  );
}

export default Movements;
