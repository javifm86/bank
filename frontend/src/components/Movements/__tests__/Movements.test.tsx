import { render, screen } from '@testing-library/react';
import Movements from '../Movements';

describe('Movements', () => {
  it('should render movements', () => {
    const mockMovements: Movements[] = [
      {
        date: '2022-01-01',
        type: 'deposit',
        amount: 100,
        balance: 100,
      },
      {
        date: '2022-01-02',
        type: 'withdraw',
        amount: 50,
        balance: 50,
      },
    ];
    const { asFragment } = render(<Movements movements={mockMovements} />);
    const rows = screen.getAllByRole('row');

    expect(asFragment()).toMatchSnapshot();

    // Header + 2 movements
    expect(rows.length).toBe(3);
  });
});
