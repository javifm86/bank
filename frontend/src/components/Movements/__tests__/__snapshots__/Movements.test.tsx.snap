// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movements should render movements 1`] = `
<DocumentFragment>
  <div
    class="-mx-4 sm:-mx-8 px-4 sm:px-8 overflow-x-auto"
  >
    <div
      class="inline-block min-w-full shadow sm:rounded-lg border-b border-gray-200 overflow-hidden"
    >
      <table
        class="min-w-full divide-y divide-gray-200"
      >
        <thead
          class="bg-gray-50"
        >
          <tr>
            <th
              class="px-5 py-3 text-xs font-semibold text-gray-500 uppercase tracking-wider text-center"
            >
              Date
            </th>
            <th
              class="px-5 py-3 text-xs font-semibold text-gray-500 uppercase tracking-wider text-right"
            >
              Amount
            </th>
            <th
              class="px-5 py-3 text-xs font-semibold text-gray-500 uppercase tracking-wider text-right"
            >
              Balance
            </th>
          </tr>
        </thead>
        <tbody
          class="bg-white divide-y divide-gray-200"
        >
          <tr>
            <td
              class="px-5 py-5 text-sm text-center"
            >
              January 1, 2022
            </td>
            <td
              class="px-5 py-5 text-sm text-right text-green-500"
            >
              +$100.00
            </td>
            <td
              class="px-5 py-5 text-sm text-right"
            >
              $100.00
            </td>
          </tr>
          <tr>
            <td
              class="px-5 py-5 text-sm text-center"
            >
              January 2, 2022
            </td>
            <td
              class="px-5 py-5 text-sm text-right text-red-500"
            >
              -$50.00
            </td>
            <td
              class="px-5 py-5 text-sm text-right"
            >
              $50.00
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
